src/coord/src/persistence.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+
+use log::{debug, error};
+use timely::progress::Antichain;
+use tokio::sync::mpsc;
+
+use dataflow_types::Update;
+use expr::GlobalId;
+use repr::Timestamp;
+use storage::{Compacter, CompacterMessage, Message, Trace, WriteAheadLogs};
+
+#[derive(Clone, Debug)]
+pub struct PersistenceConfig {
+    pub wals_path: PathBuf,
+    pub traces_path: PathBuf,
+}
+
+pub struct PersistentTables {
+    wals: WriteAheadLogs,
+    compacter_tx: mpsc::UnboundedSender<CompacterMessage>,
+    wals_path: PathBuf,
+    traces_path: PathBuf,
+    disabled: bool,
+}
+
+impl PersistentTables {
+    pub fn new(config: &PersistenceConfig) -> Option<Self> {
+        let (compacter_tx, compacter_rx) = mpsc::unbounded_channel();
+        let mut compacter = match Compacter::new(
+            compacter_rx,
+            config.traces_path.clone(),
+            config.wals_path.clone(),
+        ) {
+            Ok(compacter) => compacter,
+            Err(e) => {
+                error!(
+                    "Encountered error while trying to initialize compacter task: {}",
+                    e
+                );
+                return None;
+            }
+        };
+        tokio::spawn(async move { compacter.run().await });
```

Comment:
I am not sure this is the right place to put this comment, but recovery needs to be synchronous and needs to happen completely before the server or at least the table is made accessible for reading.

Currently, what happens is like this:
1. Insert say 100K rows
2. COUNT(*) confirms 100K insertions
2. Hit Ctrl+C to stop MZ and then restart it
3. COUNT(*) is now back to some smaller value and slowly edges back to the correct number
4. Furtheremore,  ```AS OF NOW()``` can not be used to block the query until the recovery has fully caught up.

Reply:
Do you by chance have the input file you used to do 100k inserts handy? (no worries if not I can write a script just wanted to ask)

Reply:
I generate them as SQL and pipe them through the psql client:

```
print("CREATE TABLE t1 (f1 TEXT);")
for i in range(0,100000):
   print("INSERT INTO t1 VALUES ('{}{}');".format('a' * (1024 * 32),i));
```